### PR TITLE
Add vim syntax file for MIGX IR

### PR DESCRIPTION
**Pull request by CharlieL7**
* I created a syntax highlighting file for vim for the MIGX IR
* You can add the syntax file to your syntax directory for vim or neovim to install
* Here is an example of the highlighting:
2024-02-14-191219_1276x1381_scrot

### CODE REVIEW

File: tools/syntax/migx.vim [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+syn match migx_output_type "\s->\s\zs.\{-}\ze," nextgroup=migx_output_dims
+syn match migx_output_dims "\s\zs{.\{-}}\ze,\starget_id"
+syn match migx_exec_time ":\s\zs.\{-}ms\ze"
+
+let b:current_syntax = "migx"
+
+hi def link migx_keyword NonText
+hi def link migx_ins_number Number
+hi def link migx_instruction Operator
+hi def link migx_attributes Comment
```
Comment (lakhinderwalia) [inline tools/syntax/migx.vim:27]: migx_attributes isn't always present, from what I see: example, the mul operator:
```
@13514 = reshape[dims={-1}](@13513) -> int64_type, {1}, {1}, target_id=0
@13515 = mul(@11820,@11819) -> int64_type, {1}, {1}, target_id=0
```

**Review by lakhinderwalia — APPROVED** [review]
Thanks for doing this. Very useful.

File: tools/syntax/migx.vim [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+" Current Maintainer: Charlie Lin (https://github.com/CharlieL7)
+" Previous Maintainer:
+" Last Change: 14 Feb 2024
+
+" quit when a syntax file was already loaded
+if exists("b:current_syntax")
+  finish
+endif
+
+syn keyword migx_keyword param
```
Comment (umangyadav) [inline tools/syntax/migx.vim:12]: is literal covered ?

**Review by umangyadav — APPROVED** [review]
Thanks charlie, this is useful. 

I wonder if there a way to add "go to definition" kind of jumps using ctags or other tools.

**Comment by umangyadav** [comment]
Also would it be possible to use this kind of highlighting inside vscode ? let's say i save a file as `gelu.mgx` can it be automatically highlighted ?

File: tools/syntax/migx.vim (hunk 2 above)
Comment (CharlieL7) [reply tools/syntax/migx.vim:12]: Let me make an addition for literal

**Comment by CharlieL7** [comment]
> I wonder if there a way to add "go to definition" kind of jumps using ctags or other tools.

Yes, that's possible to do with ctags in vim. I might look into it later.

> Also would it be possible to use this kind of highlighting inside vscode ? let's say i save a file as gelu.mgx can it be automatically highlighted ?

Probably, I don't use vscode however. I know how to do the auto filetype detection for vim.

> migx_attributes isn't always present, from what I see: example, the mul operator:

Good catch, latest update should fix it.

**Review by bpickrel — COMMENTED** [review]
I wonder if the boilerplate copyright header is required?

File: tools/syntax/migx.vim [hunk 3]
```diff
@@ -0,0 +1,31 @@
+" Vim syntax file
+" Language: MIGraphX Intermediate Representation
+" Current Maintainer: Charlie Lin (https://github.com/CharlieL7)
+" Previous Maintainer:
+" Last Change: 15 Feb 2024
+
```
Comment (bpickrel) [inline tools/syntax/migx.vim:6]: Is this intended to go into output files with a specific filename extension?  If so, you should describe a procedure here for registering this syntax file.  If it's just for use from the command line, then I suggest you add the following comment:

" To use from command line:  vim -u tools/syntax/migx.vim myfile

Comment (CharlieL7) [reply tools/syntax/migx.vim:6]: I'm pretty sure it's the vim tradition to have the syntax files like this. The user can decide how they want to use the file or use filetype detection.

**Pull request merged**